src/nodes.coffee

```diff
@@ -1117,50 +1232,63 @@ exports.Class = class Class extends Base
 
   # Merge the properties from a top-level object as prototypal properties
   # on the class.
-  addProperties: (node, name, o) ->
+  addProperties: (klass, name, o, node) ->
     props = node.base.properties[..]
     exprs = while assign = props.shift()
       if assign instanceof Assign
         base = assign.variable.base
-        delete assign.context
         func = assign.value
+
         if base.value is 'constructor'
-          if @ctor
-            assign.error 'cannot define more than one constructor in a class'
-          if func.bound
-            assign.error 'cannot define a constructor as a bound function'
           if func instanceof Code
-            assign = @ctor = func
+            klass.setConstructor new Access(base), func
           else
-            @externalCtor = o.classScope.freeVariable 'ctor'
-            assign = new Assign new IdentifierLiteral(@externalCtor), func
+            ctor = new IdentifierLiteral o.classScope.freeVariable 'ctor'
+            klass.setConstructor new Access(base), ctor
+            new Assign ctor, func
+        else if func instanceof Code and node in @body.expressions
+          @detectMethod klass, assign, required: true
         else
-          if assign.variable.this
-            func.static = yes
+          delete assign.context
+          accesses = if base.value == name
+            assign.variable.properties
           else
-            acc = if base.isComplex() then new Index base else new Access base
-            assign.variable = new Value(new IdentifierLiteral(name), [(new Access new PropertyName 'prototype'), acc])
-            if func instanceof Code and func.bound
-              @boundFuncs.push base
-              func.bound = no
-      assign
+            [
+              new Access new PropertyName 'prototype'
+              if base.isComplex() then new Index base else new Access base
+            ]
+          assign.variable = new Value new IdentifierLiteral(name), accesses
+          assign
+      else
+        assign
     compact exprs
 
   # Walk the body of the class, looking for prototype properties to be converted
   # and tagging static assignments.
-  walkBody: (name, o) ->
+  walkBody: (klass, name, o) ->
     @traverseChildren false, (child) =>
       cont = true
       return false if child instanceof Class
       if child instanceof Block
-        for node, i in exps = child.expressions
-          if node instanceof Assign and node.variable.looksStatic name
-            node.value.static = yes
+        for node, i in child.expressions
+          if node instanceof Assign and node.variable.base.value == name and node.value instanceof Code
```

Comment:
This no longer only detects `static` methods to ensure that `@prototype.a = ->` is captured as a class method.